### PR TITLE
Deprecate arr.device_buffer and arr.device_buffers

### PR DESCRIPTION
**Pull request by jakevdp**
Re-lands #18728, which had to be rolled back because of an incorrect test annotation.

### CODE REVIEW

File: jax/_src/test_util.py [hunk 1]
```diff
@@ -1116,13 +1116,11 @@ class BufferDonationTestCase(JaxTestCase):
   assertNotDeleted = lambda self, x: self._assertDeleted(x, False)
 
   def _assertDeleted(self, x, deleted):
-    if hasattr(x, "_arrays"):
+    if hasattr(x, "_arrays") or hasattr(x, "is_deleted"):
```
Comment (yashk2810) [inline jax/_src/test_util.py:1119]: Why do you need the `or` thing here?

Comment (jakevdp) [reply jax/_src/test_util.py:1119]: The previous logic had `if hasattr(x, "device_buffer")` and then called `device_buffer.is_deleted`. The `or` here was an attempt to replicate the previous logic.

Comment (jakevdp) [reply jax/_src/test_util.py:1119]: To be honest, I didn't understand the `elif` statement that was there before (what object would have `_arrays` but not `device_buffer`?) But my rule of thumb is to not make any assumptions when it comes to test code.

File: tests/shard_map_test.py [hunk 2]
```diff
@@ -90,6 +90,8 @@ def tearDownModule():
   xla_bridge.get_backend.cache_clear()
 
 
+@jtu.ignore_warning(category=DeprecationWarning,
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply tests/shard_map_test.py:93]: Done

File: jax/_src/test_util.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/test_util.py]: Alright, I removed the `hasattr(x, '_arrays')` case. Let's see if there are any failures.

**Review by yashk2810 — APPROVED** [review]

**Pull request merged**